File: hena/baekjoon/DfsAndBfs/14502.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

int n, m;
#define MAX 8
bool visited[MAX][MAX];
int graph[MAX][MAX];
vector<pair<int, int>> v;
int SafeAreaCount;

int dx[4] = {1, -1, 0, 0};
int dy[4] = {0, 0, 1, -1};

void init()
{
    for (int i = 0 ; i < n ; i++)
    {
        for (int j = 0 ; j < m ; j++)
        {
            visited[i][j] = false;
        }
    }
}

void CountingSafeArea()
{
    int cnt = 0;
    for (int i= 0 ; i < n ; i++)
    {
        for (int j = 0 ; j < m ; j++)
        {
            if (graph[i][j] == 0 && !visited[i][j])
                cnt++;
        }
    }
    if (SafeAreaCount < cnt)
        SafeAreaCount = cnt;
}

void BFS()
{
    queue<pair<int, int>> q;
    init();
    for (int i = 0 ; i < v.size() ; i++){
        int x = v[i].first;
        int y = v[i].second;
        q.push({x, y});
    }
    
    while (!q.empty())
    {
        int x = q.front().first;
        int y = q.front().second;
        q.pop();

        for (int i = 0 ; i < 4 ; i++)
        {
            int nx = x + dx[i];
            int ny = y + dy[i];
            if (nx < 0 || nx >= n || ny < 0  || ny >=m)
                continue;
            if (!visited[nx][ny] && graph[nx][ny] == 0)
            {
                visited[nx][ny] = true;
                q.push({nx,ny});
            }
        }
    }
    CountingSafeArea();
    
}

void recursive(int x, int y, int cnt)
{
    if (cnt == 3)
    {
        BFS();
        // print();
        return ;
    }
    for (int i = x ; i < n ; i++)
    {
        for (int j = 0 ; j < m ; j++)
        {
            if (graph[i][j] == 0)
            {
                graph[i][j] = 1;
                recursive(i, j, cnt + 1);
                graph[i][j] = 0;
            }
        }
    }
}

void find()
{
    for (int i = 0 ; i < n ; i++)
    {
        for (int j = 0 ; j < m ; j++)
        {
            if (graph[i][j] == 2)
                v.push_back({i, j});
        }
    }
    // for (int i = 0 ; i < v.size() ; i++)
    //     cout << v[i].first << v[i].second << ' ';
}
void input()
{
    cin >> n >> m;
    for (int i = 0 ; i < n ; i++)
    {
        for (int j = 0 ; j < m ; j++)
        {
            cin >>graph[i][j];
        }
    }
}
int main()
{
    input();
    find();
    recursive(0,0,0);  
    cout << SafeAreaCount;
}

File: hena/baekjoon/DfsAndBfs/14502.py
from collections import deque
import sys
input = sys.stdin.readline
dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]

def check(Map, visited, n, m):
    global Max
    cnt = 0
    for i in range(n):
        for j in range(m):
            if not visited[i][j] and Map[i][j] == 0:
                cnt += 1
    if Max < cnt:
        Max = cnt
def BFS(Map, n, m):
    visited = [[False]*m for _ in range(n)]
    global tmp
    queue = tmp
    queue = deque(queue)
    for i in queue:
        visited[i[0]][i[1]] = True

    while queue:
        x, y = queue.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if nx < 0 or nx >= n or ny < 0 or ny >= m:
                continue
            if not visited[nx][ny] and Map[nx][ny] == 0:
                queue.append([nx, ny])
                visited[nx][ny] = True
    check(Map, visited, n, m)    


def printf(Map, n, m):
    print("start")
    for i in range(n):
        for j in range(m):
            print(Map[i][j], end=" ")
        print()
    print("-end-")
def recursive(Map, n, m, d, x, y):
    global cnt
    if d == 3:
        BFS(Map, n, m)
        return
    
    for i in range(x, n):
        for j in range(m):
            if Map[i][j] == 0:
                Map[i][j] = 1
                # printf(Map, n, m)
                recursive(Map, n, m, d + 1, i, j)
                Map[i][j] = 0

def findVirus(tmp, n, m):
    for i in range(n):
        for j in range(m):
            if Map[i][j] == 2:
               tmp.append([i, j]) 

n, m = map(int, input().split())
Map = [list(map(int, input().split())) for _ in range(n)]
cnt = 0
tmp = []
Max = 0
findVirus(tmp, n, m)
recursive(Map, n, m, 0, 0, 0)

print(Max)
